blockstore/splitstore/markset_badger.go

```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
 	}
 }
 
+func (s *BadgerMarkSet) Visit(c cid.Cid) (bool, error) {
+	key := c.Hash()
+	pendKey := string(key)
+
+	checkPending := func() (bool, error) {
+		if s.pend == nil {
+			return false, errMarkSetClosed
+		}
+
+		if _, ok := s.pend[pendKey]; ok {
+			return false, nil
+		}
+
+		for _, wr := range s.writing {
+			if _, ok := wr[pendKey]; ok {
+				return false, nil
+			}
+		}
+
+		return true, nil
+	}
+
+	s.mx.RLock()
+
+	visit, err := checkPending()
+	if !visit {
+		s.mx.RUnlock()
+		return visit, err
+	}
+
+	err = s.db.View(func(txn *badger.Txn) error {
+		_, err := txn.Get(key)
+		return err
+	})
+
+	switch err {
+	case nil:
+		s.mx.RUnlock()
+		return false, nil
+
+	case badger.ErrKeyNotFound:
+		// need to upgrade the lock to exclusive in order to write; take the version count to see
+		// if there was another write while we were upgrading
+		version := s.version
+		s.mx.RUnlock()
+
+		s.mx.Lock()
```

Comment:
If we move this out of the switch statement, we can defer the unlock.

Reply:
I want to be able to do the write without holding the lock.